service/src/common/network_subgraph.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright 2023-, GraphOps and Semiotic Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+use reqwest::{header, Client, Url};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::sync::Arc;
+
+use crate::query_processor::{QueryError, UnattestedQueryResult};
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct Response<T> {
+    pub result: T,
+    pub status: i64,
+}
+
+/// Network subgraph query wrapper
+///
+/// This is Arc internally, so it can be cloned and shared between threads.
+#[derive(Debug, Clone)]
+pub struct NetworkSubgraph {
+    client: Client, // it is Arc
+    network_subgraph_url: Arc<Url>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct GraphQLQuery {
+    pub query: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub variables: Option<Value>,
+}
+
+impl NetworkSubgraph {
+    pub fn new(
+        graph_node_query_endpoint: Option<&str>,
+        deployment: Option<&str>,
+        network_subgraph_url: Option<&str>,
+    ) -> NetworkSubgraph {
+        //TODO: Check indexing status of the local network subgraph deployment
+        //if the deployment is healthy and synced, use local_network_subgraph_endpoint for network_subgraph_url
+        let _local_network_subgraph_endpoint = match (graph_node_query_endpoint, deployment)
+        {
+            (Some(_endpoint), Some(id)) =>
+            Some(Url::parse(graph_node_query_endpoint.unwrap())
+                .and_then(|u| u.join("/subgraphs/id/"))
+                .and_then(|u| u.join(id))
+                .expect("Could not parse graph node query endpoint for the network subgraph deployment")),
+            _ => None
+        };
+
+        let network_subgraph_url = network_subgraph_url
+            .map(|url| Url::parse(url).expect("Could not parse network subgraph url"));
+
+        let client = reqwest::Client::builder()
+            .user_agent("indexer-service")
+            .build()
+            .expect("Could not build a client to graph node query endpoint");
+
+        NetworkSubgraph {
+            client,
+            network_subgraph_url: Arc::new(
+                network_subgraph_url.expect("Could not query network subgraph"),
+            ),
+        }
+    }
```

Comment:
I have a hard time understanding this piece of code. I'd imagine that the arguments should be only `graph_node_query_endpoint` and `deployment` (both non optional) *or* only `network_subgraph_url` (non optional).

Reply:
My thinking around this is to track network subgraph at both the local deployment query endpoint and a remote endpoint. Although `graph_node_query_endpoint` and `deployment` should be used after we have indexing status resolver and be able to check that the deployment is synced. 
if all three inputs are provided then I think we could track both methods of querying network subgraph and use the remote endpoint as a backup when the local deployment isn't synced

Reply:
OK I see, I forgot about the backup remote endpoint bit.